### ISSUE TITLE
怎样一机多卡使用模型并行来微调Alpaca

### ISSUE THREAD

**Issue by MIkumikumi0116** [ymcui/Chinese-LLaMA-Alpaca#641]
### 提交前必须检查以下项目

- [X] 请确保使用的是仓库最新代码（git pull），一些问题已被解决和修复。
- [X] 由于相关依赖频繁更新，请确保按照[Wiki](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki)中的相关步骤执行
- [X] 我已阅读[FAQ章节](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki/常见问题)并且已在Issue中对问题进行了搜索，没有找到相似问题和解决方案
- [X] 第三方插件问题：例如[llama.cpp](https://github.com/ggerganov/llama.cpp)、[text-generation-webui](https://github.com/oobabooga/text-generation-webui)、[LlamaChat](https://github.com/alexrozanski/LlamaChat)等，同时建议到对应的项目中查找解决方案
- [X] 模型正确性检查：务必检查模型的[SHA256.md](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/SHA256.md)，模型不对的情况下无法保证效果和正常运行

### 问题类型

模型训练与精调

### 基础模型

Alpaca-Plus-13B

### 操作系统

Linux

### 详细描述问题

手头八张显卡，单张16G显存，跑微调脚本会OOM，用流水线并行或者tensor并行的话，八张卡的显存应该够，有办法吗

**Comment by MIkumikumi0116** [comment]
补充一下依赖
NVIDIA-SMI 520.61.05    Driver Version: 520.61.05    CUDA Version: 11.8

python 3.9

Package                  Version
------------------------ ------------
accelerate               0.20.3
aiohttp                  3.8.4
aiosignal                1.3.1
async-timeout            4.0.2
attrs                    23.1.0
certifi                  2023.5.7
charset-normalizer       3.1.0
cmake                    3.25.0
contourpy                1.1.0
cycler                   0.11.0
datasets                 2.13.0
deepspeed                0.9.4
dill                     0.3.6
filelock                 3.12.2
fonttools                4.40.0
frozenlist               1.3.3
fsspec                   2023.6.0
hjson                    3.1.0
huggingface-hub          0.15.1
idna                     3.4
importlib-resources      5.12.0
Jinja2                   3.1.2
joblib                   1.2.0
kiwisolver               1.4.4
lit                      15.0.7
MarkupSafe               2.1.2
matplotlib               3.7.1
mpmath                   1.2.1
multidict                6.0.4
multiprocess             0.70.14
networkx                 3.0
ninja                    1.11.1
numpy                    1.25.0
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-nvrtc-cu11   11.7.99
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96
packaging                23.1
pandas                   2.0.2
peft                     0.3.0.dev0
Pillow                   9.5.0
pip                      23.1.2
psutil                   5.9.5
py-cpuinfo               9.0.0
pyarrow                  12.0.1
pydantic                 1.10.9
pyparsing                3.1.0
python-dateutil          2.8.2
pytz                     2023.3
PyYAML                   6.0
regex                    2023.6.3
requests                 2.31.0
safetensors              0.3.1
scikit-learn             1.2.2
scipy                    1.10.1
sentencepiece            0.1.97
setuptools               67.8.0
six                      1.16.0
sklearn                  0.0.post5
sympy                    1.11.1
threadpoolctl            3.1.0
tokenizers               0.13.3
torch                    2.0.1+cu118
torchaudio               2.0.2+cu118
torchvision              0.15.2+cu118
tqdm                     4.65.0
transformers             4.28.1
triton                   2.0.0
typing_extensions        4.6.3
tzdata                   2023.3
urllib3                  2.0.3
wheel                    0.38.4
xxhash                   3.2.0
yarl                     1.9.2
zipp                     3.15.0

**Comment by MIkumikumi0116** [comment]
把ds_zero2_no_offload.json里的改成下面这样
"zero_optimization": {
        "stage": 3,
        "offload_optimizer": {
            "device":"cpu",
            "pin_memory": true
        },
        "offload_param":{
            "device":"cpu",
            "pin_memory": true
        }
    },
再加一条"zero_force_ds_cpu_optimizer": false

可以在小显卡上跑起来，当然速度会挺慢的。虽然还是没实现模型并行，不过起码能跑了。

有多张卡的话，可以把run_sft.sh里的torchrun --nnodes 1 --nproc_per_node 1，nproc_per_node 改成显卡的数量，可以用多张卡数据并行

**Comment by Naozumi520** [comment]
> 把ds_zero2_no_offload.json里的改成下面这样 "zero_optimization": { "stage": 3, "offload_optimizer": { "device":"cpu", "pin_memory": true }, "offload_param":{ "device":"cpu", "pin_memory": true } }, 再加一条"zero_force_ds_cpu_optimizer": false
> 
> 可以在小显卡上跑起来，当然速度会挺慢的。虽然还是没实现模型并行，不过起码能跑了。
> 
> 有多张卡的话，可以把run_sft.sh里的torchrun --nnodes 1 --nproc_per_node 1，nproc_per_node 改成显卡的数量，可以用多张卡数据并行

能把完整的`ds_zero2_no_offload.json`發給我嗎, 謝謝.
